pandas/core/dtypes/cast.py

```diff
@@ -1085,3 +1085,24 @@ def cast_scalar_to_array(shape, value, dtype=None):
     values.fill(fill_value)
 
     return values
+
+
+def _maybe_convert_indexer(indexer, until):
+    """
+    Convert slice, tuple, list or scalar "indexer" to 1-d array of indices,
```

Comment:
Just for completeness: recall that part of this (i.e. the ``InfoCleaner``) is  temporary, until I finish the work on ``BlockManager`` (while ``_maybe_convert_to_indexer`` will stay - it is used by the ``BlockManager`` too - as well as part of the new ``_setter``). In the long run, I expect the amount and complexity of indexing code to actually _decrease_. So I'm a bit against more hierarchy. Actually, if you find this PR too messy, I'd rather close it and just wait for the work on ``BlockManager`` (which however will take more time, so the bug will have to wait).